RecFCCeeCalorimeter/src/components/CellPositionsHCalPhiThetaSegTool.cpp

```diff
@@ -0,0 +1,211 @@
+#include "CellPositionsHCalPhiThetaSegTool.h"
+
+#include "edm4hep/CalorimeterHitCollection.h"
+#include <DDRec/DetectorData.h>
+
+using dd4hep::DetElement;
+
+DECLARE_COMPONENT(CellPositionsHCalPhiThetaSegTool)
+
+CellPositionsHCalPhiThetaSegTool::CellPositionsHCalPhiThetaSegTool(
+    const std::string &type,
+    const std::string &name,
+    const IInterface *parent)
+    : GaudiTool(type, name, parent)
+{
+  declareInterface<ICellPositionsTool>(this);
+}
+
+StatusCode CellPositionsHCalPhiThetaSegTool::initialize()
+{
+  StatusCode sc = GaudiTool::initialize();
+  if (sc.isFailure())
+    return sc;
+  m_geoSvc = service("GeoSvc");
+  if (!m_geoSvc)
+  {
+    error() << "Unable to locate Geometry service." << endmsg;
+    return StatusCode::FAILURE;
+  }
+  // get PhiTheta segmentation
+  m_segmentation = dynamic_cast<dd4hep::DDSegmentation::FCCSWGridPhiTheta_k4geo *>(
+      m_geoSvc->getDetector()->readout(m_readoutName).segmentation().segmentation());
+  if (m_segmentation == nullptr)
+  {
+    error() << "There is no phi-theta segmentation!!!!" << endmsg;
+    // return StatusCode::FAILURE;
+  }
+  // Take readout bitfield decoder from GeoSvc
+  m_decoder = m_geoSvc->getDetector()->readout(m_readoutName).idSpec().decoder();
+  m_volman = m_geoSvc->getDetector()->volumeManager();
+
+  // check if decoder contains "layer"
+  std::vector<std::string> fields;
+  for (uint itField = 0; itField < m_decoder->size(); itField++)
+  {
+    fields.push_back((*m_decoder)[itField].name());
+  }
+  auto iter = std::find(fields.begin(), fields.end(), "layer");
+  if (iter == fields.end())
+  {
+    error() << "Readout does not contain field: 'layer'" << endmsg;
```

Comment:
should we return a StatusCode::FAILURE here? or is it legitimate that the code continues running if the readout does not contain the layer field?

Reply:
added StatusCode::FAILURE since the layer info is needed for the position calculations